include/boost/container_hash/hash.hpp

```diff
@@ -307,52 +308,56 @@ namespace boost
              return seed;
         }
 
-        template <typename SizeT>
-        inline void hash_combine_impl(SizeT& seed, SizeT value)
+        template<int Bits> struct hash_combine_impl
         {
-            seed ^= value + 0x9e3779b9 + (seed<<6) + (seed>>2);
-        }
+            template <typename SizeT>
+            inline static SizeT fn(SizeT seed, SizeT value)
+            {
+                seed ^= value + 0x9e3779b9 + (seed<<6) + (seed>>2);
+                return seed;
+            }
+        };
 
-        inline void hash_combine_impl(boost::uint32_t& h1,
-                boost::uint32_t k1)
+        template<> struct hash_combine_impl<32>
         {
-            const uint32_t c1 = 0xcc9e2d51;
-            const uint32_t c2 = 0x1b873593;
-
-            k1 *= c1;
-            k1 = BOOST_FUNCTIONAL_HASH_ROTL32(k1,15);
-            k1 *= c2;
+            inline static boost::uint32_t fn(boost::uint32_t h1, boost::uint32_t k1)
+            {
+                const boost::uint32_t c1 = 0xcc9e2d51;
+                const boost::uint32_t c2 = 0x1b873593;
 
-            h1 ^= k1;
-            h1 = BOOST_FUNCTIONAL_HASH_ROTL32(h1,13);
-            h1 = h1*5+0xe6546b64;
-        }
+                k1 *= c1;
+                k1 = BOOST_FUNCTIONAL_HASH_ROTL32(k1,15);
+                k1 *= c2;
 
+                h1 ^= k1;
+                h1 = BOOST_FUNCTIONAL_HASH_ROTL32(h1,13);
+                h1 = h1*5+0xe6546b64;
 
-// Don't define 64-bit hash combine on platforms without 64 bit integers,
-// and also not for 32-bit gcc as it warns about the 64-bit constant.
-#if !defined(BOOST_NO_INT64_T) && \
-        !(defined(__GNUC__) && ULONG_MAX == 0xffffffff)
+                return h1;
+            }
+        };
 
-        inline void hash_combine_impl(boost::uint64_t& h,
-                boost::uint64_t k)
+        template<> struct hash_combine_impl<64>
         {
-            const boost::uint64_t m = UINT64_C(0xc6a4a7935bd1e995);
-            const int r = 47;
+            inline static boost::uint64_t fn(boost::uint64_t h, boost::uint64_t k)
+            {
+                const boost::uint64_t m = (boost::uint64_t(0xc6a4a793) << 32) + 0x5bd1e995;
```

Comment:
Just curious, why the switch away from `UINT64_C()` here? When I read bit manipulation code I think I still expect to see `|` used rather than `+`.